File: app/src/main/java/com/ntech/fourtop/ui/logo/LogoViewModel.java
package com.ntech.fourtop.ui.logo;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.ntech.fourtop.data.AppRepository;
import com.ntech.fourtop.network.pojo.ResponseBase;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class LogoViewModel extends ViewModel {

    private AppRepository repository;
    private CompositeDisposable disposables;
    private MutableLiveData<String> throwable;
    private MutableLiveData<String> errorMessage;
    private MutableLiveData<Integer> liveDataResponse;

    public LogoViewModel() {
        repository = AppRepository.get();
        disposables = new CompositeDisposable();
        throwable = new MutableLiveData<>();
        errorMessage = new MutableLiveData<>();
        liveDataResponse = new MutableLiveData<>();
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        disposables.clear();
    }

    public LiveData<String> getThrowable() {
        return throwable;
    }

    public LiveData<String> getErrorMessage() {
        return errorMessage;
    }

    public LiveData<Integer> getLiveDataResponse() {
        return liveDataResponse;
    }

    public void tokenRequest(String token) {
        Timber.d("Send token request");
        Disposable disposable = repository.tokenRequest(token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this::onSuccess, this::onError);
        disposables.add(disposable);
    }

    private void onSuccess(ResponseBase response) {
        int status = response.getStatus();
        Timber.d("Request status %d", status);

        switch (status) {
            case 200:
                liveDataResponse.setValue(status);
                break;
            case 403:
                errorMessage.setValue(response.getErrorMsg());
                break;
        }
    }

    private void onError(Throwable t) {
        Timber.d(t);
        throwable.setValue(t.toString());
    }
}

File: app/src/main/java/com/emikhalets/voteapp/view/logo/LogoViewModel.kt
package com.emikhalets.voteapp.view.logo

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.emikhalets.voteapp.data.AppRepository
import com.emikhalets.voteapp.data.AppRepository.Companion.get
import com.emikhalets.voteapp.network.pojo.ResponseBase
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import timber.log.Timber

class LogoViewModel : ViewModel() {
    private val repository: AppRepository?
    private val disposables: CompositeDisposable
    private val throwable: MutableLiveData<String?>
    private val errorMessage: MutableLiveData<String?>
    private val liveDataResponse: MutableLiveData<Int>
    override fun onCleared() {
        super.onCleared()
        disposables.clear()
    }

    fun getThrowable(): LiveData<String?> {
        return throwable
    }

    fun getErrorMessage(): LiveData<String?> {
        return errorMessage
    }

    fun getLiveDataResponse(): LiveData<Int> {
        return liveDataResponse
    }

    fun tokenRequest(token: String?) {
        Timber.d("Send token request")
        val disposable = repository!!.tokenRequest(token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response: ResponseBase -> this.onSuccess(response) }) { t: Throwable -> onError(t) }
        disposables.add(disposable)
    }

    private fun onSuccess(response: ResponseBase) {
        val status = response.status
        Timber.d("Check token request status %d", status)
        when (status) {
            200 -> liveDataResponse.setValue(status)
            403 -> errorMessage.setValue(response.errorMsg)
        }
    }

    private fun onError(t: Throwable) {
        Timber.d(t)
        throwable.value = t.message
    }

    init {
        repository = get()
        disposables = CompositeDisposable()
        throwable = MutableLiveData()
        errorMessage = MutableLiveData()
        liveDataResponse = MutableLiveData()
    }
}